Data link

danbryce opened the pull request.

danbryce commented on paml/paml.ttl:
```diff
@@ -355,32 +355,51 @@ paml:SampleArray rdf:type owl:Class ;
             [ rdf:type owl:Restriction ;
               owl:onProperty paml:contents ;
-              owl:allValuesFrom xsd:string ; # need to replace this with an serializable array model
+              owl:allValuesFrom paml:PrimitiveArray ; # need to replace this with an serializable array model
```
@jakebeal @rpgoldman This is the entry point to the declaration of the SampleArray contents as a PrimitiveArray.  I reused the OrderedPropertyValue to take care of indices.  OrderedPropertyValue can only hold Identified objects, and I chose to implement the contents in my test case with LiteralString objects of the form "A1", "A2", etc. to signify the identifier for aliquots.  We could also extend this to be a dict with attributes ala samples.json.

rpgoldman commented on paml/data.py:
```diff
@@ -0,0 +1,66 @@
+"""
+Functions related to data i/o in connection with a protocol execution trace.
```
Maybe add a comment at the top to point out that what you are doing here is not so much creating functions, as monkey-patching the `SampleArray` class?

rpgoldman commented on paml/paml.ttl:
```diff
@@ -355,32 +355,51 @@ paml:SampleArray rdf:type owl:Class ;
+paml:PrimitiveArray rdf:type owl:Class ;
+        rdfs:comment '''A PrimitiveArray is a one-dimensional array of primtitive types represented as strings''' ;
+        rdfs:subClassOf sbol:Identified ,
+        [ rdf:type owl:Restriction ;
```
This restriction is ill-formed.  It needs to be two restrictions: one for the `allValuesFrom` and one for the cardinality.

Also, do you really want the `rdfs:comment` here?  The comment will be stuck to the anonymous class of "things that have 0 or more `paml:elements` that are `uml:OrderedPropertyValue`s," (modulo the problem above) and not to the `paml:elements` property.

rpgoldman commented on paml/paml.ttl:
```diff
@@ -436,7 +455,7 @@ paml:SampleData rdf:type owl:Class ;
                             [ rdf:type owl:Restriction ;
```
also ill-formed

jakebeal commented on paml/data.py:
```diff
@@ -0,0 +1,66 @@
+    row_col_pairs = coordinate_rect_to_row_col_pairs("A1:H12")
+    aliquots = [f"{num2col(c+1)}{r}" for (r, c) in row_col_pairs]
+    sources = [self.name for i in range(len(aliquots))]
+    df = pd.DataFrame({
```
Since everything is going to/from dataframes, and since dataframes have a nice compact serialization form, should we just adopt dataframe serialization directly as are means of representing normal arrays? The only reason to insist on reifying the arrays into RDF is if we want to be able to perform RDF queries on them, which I think would be generally not preferred.

danbryce replied:
The solutions I found for representing CSVs or data frames were pretty clunky wrt. describing the columns using RDF.  These functions here are more of helpers to pull data into notebooks for analysis.  I think we shouldn't use this as a serialization format, unless we want to build a way to describe the tables in terms of RDF.

jakebeal replied:
I was thinking of something much simpler, actually, of just embedding a dataframe serialization as a string-valued property.

danbryce replied:
Yes, I think that could work.  However, are data frame serializations portable?  I know that csv should work ...

danbryce replied:
I have it working with pandas going to and from a string csv.

rpgoldman commented:
Could we make the data frame representation be something like a pair of encoding spec and blob (or URL) so that it's easy to extend and incorporate better representations?  E.g., the CSV format is lossy with respect to attributes, data types, etc., and string reps may be painful to store in line.  So getting something that works now, without locking us into something bad, is the priority.

jakebeal commented:
@rpgoldman That's a good idea. We do something similar in SBOL with the `Attachment` class, which I believe is one good option for implementing this.

photocyte commented:
Hi there, just chiming in if AnIML has been considered as a format for embedding/integration (referred to as a standalone [side-car file](https://en.wikipedia.org/wiki/Sidecar_file)?)

https://www.animl.org
https://en.wikipedia.org/wiki/AnIML

I believe Millipore Sigma is investing in AnIML now, so it may see some more vendor integration that it has had in the past.

danbryce commented:
I'm rebasing this branch on the latest main, and am working on the discussion points above.  I will need to force push the branch, and I'm not sure if we'll lose the discussion points.  I'm resolving them as I rebase and will push my changes when done.

danbryce commented:
Can we take this discussion of data frame serialization with an encoding spec to a new issue?  I don't think we need to solve it in order to get this PR finalized.  I think we just need to make sure that the current changes (here) don't prohibit potential solutions for serialization.

jakebeal commented:
I've added #113 for discussion

danbryce commented:
@rpgoldman @jakebeal I think this PR is ready to go.  To summarize, it includes the following changes:

- Added paml:PrimitiveArray as the paml:SampleArray.contents 
- Extend ExecutionEngine to generate a 96 well SampleArray (per above) for the EmptyContainer primitive output
- Added helper functions to create SampleArray, SampleMask, and SampleData
- Used string-based CSV format for SampleData.values, and helpers to go to/from pandas Dataframe
- Added test/test_sampledata.py to attach SampleData to a ProtocolExecution.

Errata:
- It would be nice to support both string-based CSVs for arrays (SampleData) and PrimitiveArray (SampleArray).  SampleData embeds some metadata in the CSV via the column names/types.  
- PrimitiveArray elements are OrderedPropertyValue types because they need to store both the data and its index in the array.  Currently, executing EmptyContainer results in a Primitive array of strings.  Each string could hold any string, including CSVs, as with SampleData.  The current implementation of EmptyContainer will generate a string that holds the aliquot id for each element the SampleArray (in the PrimitiveArray).
